### PR TITLE
Remove Python 2 support from C code

### PR DESCRIPTION
**Pull request by MSeifert04**
This removes handling Python 2 in the `.c` and `.h` files.

### CODE REVIEW

**Comment by pllim** [comment]
I tagged some maintainers for review (although others are free to chime in). For `utils`, I am okay with it as long as tests pass. Thanks!

**Comment by taldcroft** [comment]
@pllim - took myself off the review list as being unqualified.  I have basically no experience with the C API.

**Comment by MSeifert04** [comment]
I'm not sure what the failures on CircleCI are about:

```
=================================== FAILURES ===================================
_________________________________ test_imports _________________________________
[gw1] linux -- Python 3.5.2 /usr/bin/python3

    def test_imports():
        """
        This just imports all modules in astropy, making sure they don't have any
        dependencies that sneak through
        """
    
        from ...utils import find_current_module
    
        pkgornm = find_current_module(1).__name__.split('.')[0]
    
        if isinstance(pkgornm, str):
>           package = pkgutil.get_loader(pkgornm).load_module(pkgornm)

astropy/tests/tests/test_imports.py:30: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
<frozen importlib._bootstrap_external>:388: in _check_name_wrapper
    ???
<frozen importlib._bootstrap_external>:809: in load_module
    ???
<frozen importlib._bootstrap_external>:668: in load_module
    ???
<frozen importlib._bootstrap>:265: in _load_module_shim
    ???
<frozen importlib._bootstrap>:626: in _exec
    ???
<frozen importlib._bootstrap_external>:665: in exec_module
    ???
<frozen importlib._bootstrap>:222: in _call_with_frames_removed
    ???
astropy/__init__.py:286: in <module>
    log = _init_log()
astropy/logger.py:97: in _init_log
    log._set_defaults()
astropy/logger.py:468: in _set_defaults
    self.disable_warnings_logging()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <astropy.logger.AstropyLogger object at 0xf53795cc>

    def disable_warnings_logging(self):
        '''
            Disable logging of warnings.warn() calls
    
            Once called, any subsequent calls to ``warnings.warn()`` are no longer
            redirected to this logger.
    
            This can be re-enabled with ``enable_warnings_logging``.
            '''
        if not self.warnings_logging_enabled():
            raise LoggingError("Warnings logging has not been enabled")
        if warnings.showwarning != self._showwarning:
>           raise LoggingError("Cannot disable warnings logging: "
                               "warnings.showwarning was not set by this "
                               "logger, or has been overridden")
E           astropy.logger.LoggingError: Cannot disable warnings logging: warnings.showwarning was not set by this logger, or has been overridden

astropy/logger.py:238: LoggingError
___________________________ test_toplevel_namespace ____________________________
[gw1] linux -- Python 3.5.2 /usr/bin/python3

    def test_toplevel_namespace():
        import astropy
        d = dir(astropy)
>       assert 'os' not in d
E       AssertionError: assert 'os' not in ['Conf', 'TestRunner', 'UnsupportedPythonError', '__bibtex__', '__builtins__', '__cached__', ...]

astropy/tests/tests/test_imports.py:63: AssertionError
=============================== warnings summary ===============================
None
  Module already imported so can not be re-written: xdist
  Module already imported so can not be re-written: xdist

-- Docs: http://doc.pytest.org/en/latest/warnings.html
 2 failed, 9288 passed, 680 skipped, 62 xfailed, 5 xpassed, 2 warnings in 306.43 seconds 
/astropy_src/.eggs/astropy_helpers-2.0.1-py3.5.egg/astropy_helpers/git_helpers.py:114: UserWarning: Error running git: [Errno 2] No such file or directory: 'git'
  warnings.warn('Error running git: ' + str(e))
/usr/local/lib/python3.5/dist-packages/setuptools-28.7.1-py3.5.egg/setuptools/dist.py:331: UserWarning: Normalizing '3.0.dev' to '3.0.dev0'

./.run_docker_tests.sh returned exit code 1
```

Could someone restart that test to make sure it's not a fluke?

File: astropy/io/votable/src/tablewriter.c [hunk 1]
```diff
@@ -33,17 +33,6 @@ next_power_of_2(Py_ssize_t n)
...
 /******************************************************************************
- * Python version compatibility macros
- ******************************************************************************/
-#if PY_MAJOR_VERSION >= 3
-#  define IS_PY3K
-#endif
-
-#ifndef Py_TYPE
-#  define Py_TYPE(o) ((o)->ob_type)
-#endif
```
Comment (saimn) [inline astropy/io/votable/src/tablewriter.c]: This macro is always defined for PY3 ?

Comment (MSeifert04) [reply astropy/io/votable/src/tablewriter.c]: Yes, it's defined in Python2 >= 2.7 and Python3 >= 3.1.

**Comment by nden** [comment]
@MSeifert04 he changes in `projections.c.templ` look good. Did you try recreating `projections.c`? It can be part of this PR.

**Comment by MSeifert04** [comment]
@nden I'm not sure what you mean. The `projections.c` file isn't tracked by git so recreating it doesn't make it show up on the diff.

**Comment by nden** [comment]
@MSeifert04 Sorry, just shows my limited understanding of the build process. There's a comment in `porjections.c.templ` saying `projections.c` is auto-generated by `erfa_generator.py`. Now that I looked at setup_package.py it seems the comment is incorrect. Or am I confused again?

The wcs changes look OK to me.

**Comment by MSeifert04** [comment]
> Now that I looked at setup_package.py it seems the comment is incorrect. Or am I confused again?

I really don't know how exactly this is build but I tend to agree: it seems like it's just a plain jinja-template.

**Comment by MSeifert04** [comment]
Seems like the only touched subpackage still awaiting approval is `erfa`. 

Otherwise it seems ready to "go in"?

**Comment by pllim** [comment]
Huh, there are some warnings in Sphinx build and they don't look like time-outs:
```
/home/travis/build/astropy/astropy/docs/generated/examples/io/plot_fits-image.rst:12: WARNING: py:obj reference target not found: matplotlib.pyplot

/home/travis/build/astropy/astropy/docs/generated/examples/io/split-jpeg-to-fits.rst:13: WARNING: py:obj reference target not found: matplotlib.pyplot

/home/travis/build/astropy/astropy/docs/visualization/wcsaxes/index.rst:42: WARNING: py:mod reference target not found: matplotlib.pyplot
```

**Comment by MSeifert04** [comment]
@pllim The newest mpl release broke our CI: #6695

**Comment by pllim** [comment]
Approvals obtained and test failures understood (and unrelated), so merging. Thanks!

**Pull request merged**